src/keystore.cpp

```diff
@@ -399,6 +424,20 @@ std::optional<libzcash::UFVKId> CBasicKeyStore::GetUFVKIdForViewingKey(const lib
     return result;
 }
 
+std::optional<std::pair<libzcash::UFVKId, std::optional<libzcash::diversifier_index_t>>>
+FindUFVKId::operator()(const libzcash::OrchardRawAddress& orchardAddr) const {
+    for (const auto& [k, v] : keystore.mapUnifiedFullViewingKeys) {
+        auto fvk = v.GetOrchardKey();
+        if (fvk.has_value()) {
+            auto d_idx = fvk.value().ToIncomingViewingKey().DecryptDiversifier(orchardAddr);
+            if (d_idx.has_value()) {
+                return std::make_pair(k, d_idx);
+            }
+        }
+    }
+    return std::nullopt;
+}
+
 std::optional<std::pair<libzcash::UFVKId, std::optional<libzcash::diversifier_index_t>>>
 FindUFVKId::operator()(const libzcash::SaplingPaymentAddress& saplingAddr) const {
```

Comment:
Shouldn't this also decrypt the diversifier in the same way as Orchard? I think we only got away with not doing so because we were not generating Sapling diversified addresses, but now we are.

Reply:
The reason for the difference here is that for Sapling we have direct access to the `addr -> ivk` map, which is updated whenever we see a new diversified address (via note detection or address generation), and thus can be relied on. For Orchard, the keystore is over on the Rust side inside the Orchard wallet, so we don't have access to it from within the C++ keystore API, and instead need to trial-decrypt the Orchard receiver with every UFVK in the C++ keystore in order to figure out which one is which.

Reply:
I was anticipating that we would delegate the `OrchardRawAddress -> std::optional<OrchardIncomingViewingKey>` lookup to the Rust side of the Orchard wallet, and then just look up the ufvkid from `mapOrchardKeyUnified`; wouldn't that be more consistent with what we're doing for Sapling?

Reply:
We can't do that delegation for the reason I gave above: `CKeyStore` has no knowledge of the Orchard wallet, only `CWallet` does. We'd either need to break the Orchard wallet into two separately-managed Rust handles, one for each layer, or store the Orchard wallet in the `CKeyStore` layer and "pretend" it has no wallet functionality. But neither way is how you built it, so I gook that as meaning you intended for the approach I took (which IIRC was already in the branch you handed to me).

Reply:
Hm. I'm not sure what the most correct thing to do here is. I don't think it's correct to split the Orchard wallet into separately-managed handles. Another possibility is that `FindUFVKId` could be moved out of the keystore functionality and into the wallet functionality, and we could bridge the gap that way; trial decryption here definitely means that we behave differently for Orchard addresses than we do for Sapling; it would mean that you could use `FindUFVKId` to determine the UFVK corresponding to an address that was never generated by the wallet, and that is inconsistent with the rest of the address-based functionality that we provide. I'd rather keep our semantics consistent, that we only allow the user to ask questions of the wallet using addresses that were generated by the wallet, to avoid potential corner cases.

Reply:
In our pairing we decided on an approach which I've written up in #5575. In short, here we follow @daira's suggestion and trial-decrypt for Sapling against known UFVKs as well.